core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java

```diff
@@ -273,15 +277,10 @@ private boolean collides(TypeCodec<?> newCodec, TypeCodec<?> oldCodec) {
       }
     }
 
-    if (value instanceof TupleValue) {
-      return uncheckedCast(codecFor(((TupleValue) value).getType(), TupleValue.class));
-    } else if (value instanceof UdtValue) {
-      return uncheckedCast(codecFor(((UdtValue) value).getType(), UdtValue.class));
-    }
-
-    GenericType<?> javaType = inspectType(value, null);
+    DataType cqlType = inferCqlTypeFromValue(value);
```

Comment:
We need to infer a CQL type here, if possible, before calling `getCachedCodec` because once inside this method the original value is lost, and it's impossible to guess the CQL type after that.